src/cls/hash/server.cc

```diff
@@ -0,0 +1,399 @@
+  // CEPH_OSD_OP_FLAG_BYPASS_CLEAN_CACHE ???
+  uint32_t read_flags = CEPH_OSD_OP_FLAG_FADVISE_NOCACHE;
+  int ofs = 0, len = 0;
+  ceph::buffer::list bl;
+  int ret = cls_cxx_read2(hctx, ofs, len, &bl, read_flags);
+  if (ret < 0) {
+    CLS_LOG(1, "%s:: failed cls_cxx_read2() ret=%d (%s)",
+            __func__, ret, cpp_strerror(-ret).c_str());
+    return ret;
+  }
+
+  ceph::crypto::SHA256 hmac;
+  for (const auto& bptr : bl.buffers()) {
+    hmac.Update((const unsigned char *)bptr.c_str(), bptr.length());
+  }
+
+  uint8_t hash[CEPH_CRYPTO_HMACSHA256_DIGESTSIZE];
+  hmac.Final(hash);
+  out->append((const char *)hash, sizeof(hash));
+
+  return 0;
+}
+
+//------------------------------------------------------------------------------
+static inline bool multi_part_support(int32_t hash_type)
+{
+  return (hash_type == HASH_BLAKE3 && g_blake_lib_multi_part_support);
+}
+
+//------------------------------------------------------------------------------
+static int hash_data(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "hash_data main function");
+  cls_hash_op op;
+  try {
+    auto p = in->cbegin();
+    decode(op, p);
+  } catch (const buffer::error&) {
+    CLS_LOG(0, "ERROR: %s: failed to decode input", __func__);
+    return -EINVAL;
+  }
+
+  if (unlikely(!multi_part_support(op.hash_type) && !op.flags.is_single_part())) {
+    // we don't know how to serialize hash state so only single part
+    // operations are supported
+    CLS_LOG(0, "ERR: %s: hash_type=%d supports only single part hash",
+            __func__, op.hash_type);
+    return -EOPNOTSUPP;
+  }
+
+  switch (op.hash_type) {
+  case HASH_MD5:
+    return md5_hash(hctx, op, out);
+  case HASH_SHA256:
+    return sha256_hash(hctx, op, out);
+  case HASH_BLAKE3:
+    return blake3_hash(hctx, op, out);
+  default:
+    CLS_LOG(0, "ERR: %s: unexpcted hash_type (%u)", __func__, op.hash_type);
+    return -EINVAL;
+  }
+}
+
+//------------------------------------------------------------------------------
+static bool check_blake3_lib_multi_part_support()
+{
+  // The BLAKE3-CLS understands the internal data-structures holding the hash state
+  // (blake3_hasher and blake3_chunk_state)
+  // The embedded version used by ceph is 1.5.0 (from Sep 21, 2023)
+  // Current version in https://github.com/BLAKE3-team/BLAKE3 (June-19-2025)
+  // is 1.8.2.
+  // Blake-CLS will work with any version up-to 1.8.2
+  // The data-structures holding the hash state were not changed since version 1.0.0
+  // so it is very unlikely they will be changed in the future, but we still lock
+  // to the last known version (1.8.2 ) to be on the safe side
+  constexpr std::string_view ver(BLAKE3_VERSION_STRING);
+  constexpr std::string_view max_ver("1.8.2");
+  if (ver > max_ver) {
+    CLS_LOG(0, "Disable BLAKE3 multipart support: wrong lib version (%s)",
+            BLAKE3_VERSION_STRING);
+    return false;
+  }
+
+  // make sure the defined macros were not changed
```

Comment:
i think you can change everything in this function to be static.
this should break compilation if anything changes in the blake3 state which is incompatible with the current code.
i think you can leave out the version number check - sice change in version does not mean necessarily break in the code.
maybe emit an ifdef compilation warning based on the version change?

Reply:
I got another procedure doing static checks below 
I need to verify the lib version in case they manage to sneak a change my checks are unable to capture (there are 5 unused bytes after flags in blake3_chunk_state)